### ISSUE TITLE
Qwen 3.x series support

### ISSUE THREAD

**Issue by imkebe** [airockchip/rknn-llm#251]

**Comment by codezjx** [comment]
+1, 转换 [Qwen3-0.6B](https://huggingface.co/Qwen/Qwen3-0.6B) 的时候报错：
```
INFO: rkllm-toolkit version: 1.2.0
ERROR: Can not load QWenLMHeadModel!
Load model failed!
```

**Comment by KevinYoung-Kw** [comment]
+1 Need Qwen 3.x series support

**Comment by waydong** [comment]
> **EDIT** - used the wrong json file without q_norm and k_norm declared.
> 
> You can convert custom models now, but for some reason the q and k norm layers aren't supported, even though they're RMS, and almost identical to Gemma3's attention layer norm:
> 
> INFO: rkllm-toolkit version: 1.2.0
> INFO: Please check if the matched structure of the custom model is correct:
> INFO: ######################################################################
> INFO: TOKEN_EMBD -> model.embed_tokens
> INFO: ATTN_Q -> model.layers.{bid}.self_attn.q_proj
> INFO: ATTN_V -> model.layers.{bid}.self_attn.v_proj
> INFO: ATTN_OUT -> model.layers.{bid}.self_attn.o_proj
> INFO: ATTN_Q_NORM -> model.layers.{bid}.self_attn.q_norm
> INFO: ATTN_K_NORM -> model.layers.{bid}.self_attn.k_norm
> INFO: FFN_GATE -> model.layers.{bid}.mlp.gate_proj
> INFO: FFN_UP -> model.layers.{bid}.mlp.up_proj
> INFO: FFN_DOWN -> model.layers.{bid}.mlp.down_proj
> INFO: ATTN_NORM -> model.layers.{bid}.input_layernorm
> INFO: ATTN_POST_NORM -> model.layers.{bid}.post_attention_layernorm
> INFO: OUTPUT_NORM -> model.norm
> INFO: OUTPUT -> lm_head
> INFO: ######################################################################
> Building model: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 427/427 [00:05<00:00, 78.18it/s]
> INFO: Setting chat_template to "<|im_start|>user\n[content]<|im_end|>\n<|im_start|>assistant\n"
> INFO: Setting token_id of eos to 151645
> INFO: Setting token_id of pad to 151643
> INFO: Setting token_id of bos to 151643
> INFO: Setting add_bos_token to False
> Converting model: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 311/311 [00:00<00:00, 2276489.61it/s]
> ERROR: Found unsupported layer: model.layers.0.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.0.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.1.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.1.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.2.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.2.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.3.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.3.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.4.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.4.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.5.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.5.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.6.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.6.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.7.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.7.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.8.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.8.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.9.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.9.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.10.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.10.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.11.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.11.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.12.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.12.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.13.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.13.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.14.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.14.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.15.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.15.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.16.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.16.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.17.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.17.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.18.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.18.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.19.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.19.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.20.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.20.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.21.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.21.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.22.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.22.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.23.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.23.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.24.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.24.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.25.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.25.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.26.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.26.self_attn.k_norm.weight!
> ERROR: Found unsupported layer: model.layers.27.self_attn.q_norm.weight!
> ERROR: Found unsupported layer: model.layers.27.self_attn.k_norm.weight!
> INFO: Setting max_context_limit to 16384
> INFO: Exporting the model, please wait ....
> [=================================================>] 597/597 (100%)
> INFO: Model has been saved to Qwen3-0.6B/Qwen3-0.6B.rkllm!
> For a custom model, download the manual from the "doc" section of this repo, and it will show you how to map custom values. For Qwen3, it is (I believe):
> 
> {
>     "BLOCKNAME": "Qwen3DecoderLayer",
>     "TOKEN_EMBD": "embed_tokens",
>     "ATTN_NORM": "input_layernorm",
>     "ATTN_Q": "self_attn.q_proj",
>     "ATTN_Q_NORM": "self_attn.q_norm",
>     "ATTN_K_NORM": "self_attn.k_norm",
>     "ATTN_V": "self_attn.v_proj",
>     "ATTN_QKV": "",
>     "ATTN_KV": "",
>     "KV_CONTINUOUS": "false",
>     "ATTN_OUT": "self_attn.o_proj",
>     "ATTN_POST_NORM": "post_attention_layernorm",
>     "FFN_NORM": "",
>     "FFN_UP": "mlp.up_proj",
>     "FFN_GATE": "mlp.gate_proj",
>     "ACT_TYPE": "silu",
>     "FFN_DOWN": "mlp.down_proj",
>     "FFN_POST_NORM": "",
>     "OUTPUT_NORM": "norm",
>     "OUTPUT": "lm_head"
> }
> Here is model structure:
> 
> Qwen3ForCausalLM(
>   (model): Qwen3Model(
>     (embed_tokens): Embedding(151936, 1024)
>     (layers): ModuleList(
>       (0-27): 28 x Qwen3DecoderLayer(
>         (self_attn): Qwen3Attention(
>           (q_proj): Linear(in_features=1024, out_features=2048, bias=False)
>           (k_proj): Linear(in_features=1024, out_features=1024, bias=False)
>           (v_proj): Linear(in_features=1024, out_features=1024, bias=False)
>           (o_proj): Linear(in_features=2048, out_features=1024, bias=False)
>           (q_norm): Qwen3RMSNorm((128,), eps=1e-06)
>           (k_norm): Qwen3RMSNorm((128,), eps=1e-06)
>         )
>         (mlp): Qwen3MLP(
>           (gate_proj): Linear(in_features=1024, out_features=3072, bias=False)
>           (up_proj): Linear(in_features=1024, out_features=3072, bias=False)
>           (down_proj): Linear(in_features=3072, out_features=1024, bias=False)
>           (act_fn): SiLU()
>         )
>         (input_layernorm): Qwen3RMSNorm((1024,), eps=1e-06)
>         (post_attention_layernorm): Qwen3RMSNorm((1024,), eps=1e-06)
>       )
>     )
>     (norm): Qwen3RMSNorm((1024,), eps=1e-06)
>     (rotary_emb): Qwen3RotaryEmbedding()
>   )
>   (lm_head): Linear(in_features=1024, out_features=151936, bias=False)
> )
> And the RMSNorm layer, which works perfectly fine in other parts of the model:
> 
> @use_kernel_forward_from_hub("RMSNorm")
> class Qwen3RMSNorm(nn.Module):
>     def __init__(self, hidden_size, eps=1e-6):
>         """
>         Qwen3RMSNorm is equivalent to T5LayerNorm
>         """
>         super().__init__()
>         self.weight = nn.Parameter(torch.ones(hidden_size))
>         self.variance_epsilon = eps
> 
>     def forward(self, hidden_states):
>         input_dtype = hidden_states.dtype
>         hidden_states = hidden_states.to(torch.float32)
>         variance = hidden_states.pow(2).mean(-1, keepdim=True)
>         hidden_states = hidden_states * torch.rsqrt(variance + self.variance_epsilon)
>         return self.weight * hidden_states.to(input_dtype)
> 
>     def extra_repr(self):
>         return f"{tuple(self.weight.shape)}, eps={self.variance_epsilon}"
> And the relevant sections in the init and forward portions of the attention block, self_attn.
> 
> init:
> 
>         self.q_norm = Qwen3RMSNorm(self.head_dim, eps=config.rms_norm_eps)  # unlike olmo, only on the head dim!
>         self.k_norm = Qwen3RMSNorm(self.head_dim, eps=config.rms_norm_eps)  # thus post q_norm does not need reshape
> forward:
> 
>         input_shape = hidden_states.shape[:-1]
>         hidden_shape = (*input_shape, -1, self.head_dim)
> 
>         query_states = self.q_norm(self.q_proj(hidden_states).view(hidden_shape)).transpose(1, 2)
>         key_states = self.k_norm(self.k_proj(hidden_states).view(hidden_shape)).transpose(1, 2)
> Not entirely sure why it's failing. I know that the manual says RMS only for layernorm, but that's what this is, and it's not even nominal. That is very clearly a root mean square norm layer.
> 
> [@waydong](https://github.com/waydong) - any advice for getting this conversion to work? I've even tried it as a temporary buffer, but that is not in the state dict and thus can't get written to any RKLLM layers. Even though it converts without throwing errors, it attempts to initialize q weights as 1024, 1024 instead of 1024, 2048. Maybe it's because the normalization only happens on the head dim?

The custom structure does not currently support q_norm and k_norm, but we are working on adding support for them.

**Comment by c0zaut** [comment]
Just saw the new version! Thank you, @yhcvb !!! Gonna test in a bit

**Comment by zhjygit** [comment]
之前的1.1.4版本rkllm-toolkit不支持qwen3的rkllm模型转换。
现在的1.2.1版本支持如何？
我的转换代码是：
```
      from rkllm.api import RKLLM
      from datasets import load_dataset
      from transformers import AutoTokenizer
      from tqdm import tqdm
      import torch
      from torch import nn
      import os
      # os.environ['CUDA_VISIBLE_DEVICES']='1'
      
      '''
      https://huggingface.co/Qwen/Qwen-1_8B-Chat
      从上面网址中下载Qwen模型
      '''
      
      #modelpath = '/home/zhjy/rknn-llm/model/models/LLM-Research//Meta-Llama-3-8B'
      modelpath = '/home/zhjy/rknn-llm-1.2.1/model/qwen3-1.7b'
      
      # modelpath = "./path/to/Qwen-1.8B-F16.gguf"
      llm = RKLLM()
      
      # Load model
      # Use 'export CUDA_VISIBLE_DEVICES=2' to specify GPU device
      # options ['cpu', 'cuda']
      #print('ret test')
      ret = llm.load_huggingface(model=modelpath, model_lora = None, device='cpu')
      #ret = llm.load_gguf(model = modelpath)
      if ret != 0:
          print('Load model failed!')
          exit(ret)
      
      # Build model
      #dataset = "./data_quant.json"
      dataset=None
      # Json file format, please note to add prompt in the input，like this:
      # [{"input":"Human: 你好！\nAssistant: ", "target": "你好！我是人工智能助手KK！"},...]
      
      qparams = None
      # qparams = 'gdq.qparams' # Use extra_qparams
      ret = llm.build(do_quantization=True, optimization_level=1, quantized_dtype='w8a8',
                      quantized_algorithm='normal', target_platform='rk3588', num_npu_core=3, extra_qparams=qparams, dataset=dataset)
      
      if ret != 0:
          print('Build model failed!')
          exit(ret)
      
      # Evaluate Accuracy
      def eval_wikitext(llm):
          seqlen = 512
          tokenizer = AutoTokenizer.from_pretrained(
              modelpath, trust_remote_code=True)
          # Dataset download link:
          # https://huggingface.co/datasets/Salesforce/wikitext/tree/main/wikitext-2-raw-v1
          testenc = load_dataset(
              "parquet", data_files='./wikitext/wikitext-2-raw-1/test-00000-of-00001.parquet', split='train')
          testenc = tokenizer("\n\n".join(
              testenc['text']), return_tensors="pt").input_ids
          nsamples = testenc.numel() // seqlen
          nlls = []
          for i in tqdm(range(nsamples), desc="eval_wikitext: "):
              batch = testenc[:, (i * seqlen): ((i + 1) * seqlen)]
              inputs = {"input_ids": batch}
              lm_logits = llm.get_logits(inputs)
              if lm_logits is None:
                  print("get logits failed!")
                  return
              shift_logits = lm_logits[:, :-1, :]
              shift_labels = batch[:, 1:].to(lm_logits.device)
              loss_fct = nn.CrossEntropyLoss().to(lm_logits.device)
              loss = loss_fct(
                  shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
              neg_log_likelihood = loss.float() * seqlen
              nlls.append(neg_log_likelihood)
          ppl = torch.exp(torch.stack(nlls).sum() / (nsamples * seqlen))
          print(f'wikitext-2-raw-1-test ppl: {round(ppl.item(), 2)}')
      
      # eval_wikitext(llm)
      
      
      # Chat with model
      messages = "<|im_start|>system You are a helpful assistant.<|im_end|><|im_start|>user你好！\n<|im_end|><|im_start|>assistant"
      kwargs = {"max_length": 128, "top_k": 1, "top_p": 0.8,
                "temperature": 0.8, "do_sample": True, "repetition_penalty": 1.1}
      # print(llm.chat_model(messages, kwargs))
      
      
      # Export rkllm model
      ret = llm.export_rkllm("./qwen3-1.7b-tools.rkllm")
      if ret != 0:
          print('Export model failed!')
          exit(ret)
      
```

1.2.1版本已经支持Qwen3，按您提供的代码可以正常转换。
